scalding-core/src/main/scala/com/twitter/scalding/Mode.scala

```diff
@@ -103,6 +103,12 @@ trait Mode extends java.io.Serializable {
    * Make the Execution.Writer for this platform
    */
   def newWriter(): Execution.Writer
+}
+
+/**
+ * Any Mode running on cascading extends CascadingMode
+ */
+trait CascadingMode extends Mode {
```

Comment:
with this, `Mode` is really just an Execution.Writer factory (the thing that holds the mutable state while an Execution is running).

This should be source compatible with everything we have now since Mode was never sealed, so case matching always had to deal with an open hierarchy.

With this, we move the cascading specific methods onto the CascadingMode. The goal is to eventually remove cascading from being required to use Execution + TypedPipe.